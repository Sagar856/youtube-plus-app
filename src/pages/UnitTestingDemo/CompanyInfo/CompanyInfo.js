import React, { useState } from 'react';
import { PropTypes } from 'prop-types';

const CompanyInfo = (props) => {
  const [countryName, setCountryName] = useState('USA');

  return (
    <div>
      <h2>CompanyInfo | Testing Props, Placerholder Text, Events, States, Snapshots</h2>
      <h3 data-testid="companyName">Cognizant</h3>

      <hr />
      <p data-testid="foundedYearElement" style={{ color: 'rgb(0, 255, 0)' }}>
        Founded in: {props.foundedYear}
      </p>
      <p data-testid="employeeCountElement">Employee Count : {props.employeeCount}</p>
      <input
        type="text"
        placeholder="Enter Your Country Name"
        value={countryName}
        onChange={(event) => {
          setCountryName(event.target.value);
        }}
      />
      <p data-testid="visitCognizant">Visit Cognizant {countryName} Website</p>
      <hr />
      <button className="btn btn-primary" data-testid="classNameElement">
        Join Cognizant
      </button>
    </div>
  );
};

CompanyInfo.propTypes = {
  foundedYear: PropTypes.string,
  employeeCount: PropTypes.string
};

export default CompanyInfo;
