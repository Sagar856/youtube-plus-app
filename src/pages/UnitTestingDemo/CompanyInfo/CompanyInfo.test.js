import { fireEvent, render, screen } from '@testing-library/react';
import CompanyInfo from './CompanyInfo';
import renderer from 'react-test-renderer';

describe('CompanyInfo', () => {
  it('should have companyName Cognizant', () => {
    render(<CompanyInfo />);
    const companyNameElement = screen.getByTestId('companyName');
    console.log(companyNameElement.textContent);
    expect(companyNameElement.textContent).toBe('Cognizant');
  });

  // testing Props
  it('receives founderYear prop and dispaly in JSX', () => {
    // when you're testing prop, render the comp with necessary props
    render(<CompanyInfo foundedYear="2006" />);
    const foundedYearElement = screen.getByTestId('foundedYearElement');
    expect(foundedYearElement).toHaveTextContent('Founded in: 2006');
  });

  // TODO: test weather comp is receiving a prop 'employeeCount'
  it('receives employeeCount prop and dispaly in JSX', () => {
    render(<CompanyInfo employeeCount="4 lac" />);
    const employeeCountElement = screen.getByTestId('employeeCountElement');
    expect(employeeCountElement).toHaveTextContent('Employee Count : 4 lac');
  });

  // Form input with placeholder
  it("has an input with 'Enter Your Country Name' as placeholder text", () => {
    render(<CompanyInfo />);
    expect(screen.getAllByPlaceholderText('Enter Your Country Name')).toBeTruthy();
  });

  // Inline styles testing
  it('has foundedYear is with text color green--inline style', () => {
    render(<CompanyInfo foundedYear="2006" />);
    const foundedYearElement = screen.getByTestId('foundedYearElement');
    expect(foundedYearElement).toHaveStyle('color : rgb(0, 255, 0)');
  });

  // TODO: test weather button has 'btn-primary' class
  it("has css class 'btn-primary'", () => {
    render(<CompanyInfo />);
    const classNameElement = screen.getByTestId('classNameElement');
    expect(classNameElement).toHaveClass('btn-primary');

    const buttonTextElement = screen.getByTestId('classNameElement');
    expect(classNameElement).toHaveTextContent('Join Cognizant');
  });

  // Events and states
  it('displays countryName by default and updates UI onChange of input', () => {
    render(<CompanyInfo />);
    //find the input element
    const countryNameInput = screen.getByPlaceholderText('Enter Your Country Name');
    //Now checking wheather the onChange event handler is working or not
    // let's trigger the event -- thru program
    const mockEventObj = {
      // preparing the mock event obj with target value being china
      target: {
        value: 'Japan'
      }
    };
    fireEvent.change(countryNameInput, mockEventObj);
    //finally checking if the input element is showing the change
    expect(countryNameInput.value).toBe('Japan');

    // Have expectation
    expect(screen.getByTestId('visitCognizant')).toHaveTextContent('Visit Cognizant Japan Website');
  });

  // Snapshot testing
  it('has right snapshot with all requirements completed', () => {
    // to take snapshot we need react-test-renderer.  // npm i react-test-renderer
    // taking snapshot and also converting into JSON
    //[RECOMMENDED]: take snapshot with all possible props as well as props children
    const snapshotInJson = renderer.create(<CompanyInfo foundedYear="2006" />).toJSON();
    // let assert with a matchrer toMatchSnapshot()
    expect(snapshotInJson).toMatchSnapshot();
  });
});
