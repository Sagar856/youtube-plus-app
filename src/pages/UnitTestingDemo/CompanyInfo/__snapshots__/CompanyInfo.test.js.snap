// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyInfo has right snapshot with all requirements completed 1`] = `
<div>
  <h2>
    CompanyInfo | Testing Props, Placerholder Text, Events, States, Snapshots
  </h2>
  <h3
    data-testid="companyName"
  >
    Cognizant
  </h3>
  <hr />
  <p
    data-testid="foundedYearElement"
    style={
      Object {
        "color": "rgb(0, 255, 0)",
      }
    }
  >
    Founded in: 
    2006
  </p>
  <p
    data-testid="employeeCountElement"
  >
    Employee Count : 
  </p>
  <input
    onChange={[Function]}
    placeholder="Enter Your Country Name"
    type="text"
    value="USA"
  />
  <p
    data-testid="visitCognizant"
  >
    Visit Cognizant 
    USA
     Website
  </p>
  <hr />
  <button
    className="btn btn-primary"
    data-testid="classNameElement"
  >
    Join Cognizant
  </button>
</div>
`;
